ffi-proc-macros/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{quote, quote_spanned};
+use syn::parse::{Parse, ParseStream};
+use syn::spanned::Spanned;
+use syn::{
+    parse_macro_input, Error, Fields, FieldsNamed, FieldsUnnamed, Ident, ItemStruct, LitBool,
+    Result, Token, Type,
+};
+
+struct HandleDescriptorParams {
+    target: Type,
+    mutable: bool,
+    sized: bool,
+}
+
+impl Parse for HandleDescriptorParams {
+    fn parse(input: ParseStream) -> Result<Self> {
+        // Parse the "target=..." field (always first)
+        let key: Ident = input.parse()?;
+        if let "target" = key.to_string().as_str() {
+            let _: Token![=] = input.parse()?;
+        } else {
+            return Err(Error::new(key.span(), "expected `target` field"));
+        }
+        let target: Type = input.parse()?;
+        let is_trait = match target {
+            Type::Path(_) => false,
+            Type::TraitObject(_) => true,
+            _ => {
+                return Err(Error::new(
+                    target.span(),
+                    "expected a type or a trait object",
+                ))
+            }
+        };
+
+        // Parse the remaining fields (unordered)
+        let mut mutable = None;
+        let mut sized = None;
+        while !input.is_empty() {
+            let _: Token![,] = input.parse()?;
+            let ident: Ident = input.parse()?;
+            let _: Token![=] = input.parse()?;
+            let value: LitBool = input.parse()?;
+            match ident.to_string().as_str() {
+                "mutable" if mutable.is_none() => {
+                    mutable = Some(value.value());
+                }
+                "sized" if sized.is_none() => {
+                    if value.value() && is_trait {
+                        return Err(Error::new(value.span(), "trait objects are never Sized"));
+                    }
+                    sized = Some(value.value());
+                }
+                _ => {
+                    return Err(Error::new(ident.span(), "unknown or duplicated field"));
+                }
+            }
+        }
+        let mutable = mutable.ok_or_else(|| input.error("missing `mutable` field"))?;
+        if sized.is_none() && is_trait {
+            sized = Some(false);
+        }
+        let sized = sized.ok_or_else(|| input.error("missing `sized` field"))?;
```

Comment:
can't we just `unwrap` here, we just ensured it's not `None` above

Reply:
L63 only applies to trait targets (for which `sized` is optional); struct targets still require it, since I couldn't figure out a way to tell at macro expansion time whether the target struct is `Sized` or not...

Reply:
Ahh right, and I guess we could maybe have zero sized structs (although unlikely). Okay, makes sense then.